### PR TITLE
search: add separate flag for formulae

### PR DESCRIPTION
**Pull request by vidusheeamoli**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR attempts to solve issue #6975. 

Right now, this PR has a lot of scope for improvement, but I am determined to get this done right and merged. As discussed in the above issue I have added a separate `--formulae` flag in brew search and this is what brew search looks like now:

```
brew search					#lists both formulae and casks
brew search --formulae		#lists only formulae
brew search --casks 		#lists only casks
```

I have tested these changes locally, and they seem to work so far.

In this version of implementation, `brew search` first lists all cask names(sorted) followed by sorted formula names. Is this correct, or should I mix all cask and formula names in a list and then print this list in a sorted manner (this would mix all cask and formula names).

Thank you, @issyl0 for telling me such a convenient way to test out my changes locally -- it saved me a lot of time! Please take a look and let me know. 😃

Thanks! 🚀

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
Great work here @vidusheeamoli, you're off to a fantastic start! I think it'd be great to get this PR to a point where `brew search --formulae chrome` works. I'm not sure that having just `brew search` by default list casks is super useful (and it may break other tools) but having the `brew search --formulae` output just formulae does feel appropriate.

Additionally, CI is failing here so when you make your next changes run `brew man` and commit the changes there.

Thanks again!

**Comment by vidusheeamoli** [comment]
Thank you for the feedback @MikeMcQuaid! 😃 

> I'm not sure that having just `brew search` by default list casks is super useful (and it may break other tools) but having the `brew search --formulae` output just formulae does feel appropriate.

Forgive me for the silly question, but does this mean that `brew search` and `brew search --formulae` will both list only formulae? Wouldn't it be confusing for users to have 2 different commands do the exact same job? 

> Additionally, CI is failing here so when you make your next changes run `brew man` and commit the changes there.

Sure thing! I will also be adding tests for my changes.  :)

**Comment by MikeMcQuaid** [comment]
> Forgive me for the silly question, but does this mean that `brew search` and `brew search --formulae` will both list only formulae?

Yep!

> Wouldn't it be confusing for users to have 2 different commands do the exact same job?

Perhaps but I think if `--formulae` does other things (e.g. when passed arguments) then it makes sense to be consistent.



> On thinking about it, i've realised that this implementation will give more clarity as compared to the current scenario. `brew search` -- by default will list only formulae, but having the `--formulae` flag also provides more consistency to the users (since there is already a flag for casks). Hope I'm thinking in the right direction here.

Exactly!

> Sure thing! I will also be adding tests for my changes. :)

Cool! Tests aren't required but if you do add them don't make more `:integration` tests as they are slow (so we keep the numbers down).

**Comment by vidusheeamoli** [comment]
Hello @MikeMcQuaid!

I've incorporated all the discussed changes; and here is what brew search looks like now:

```
brew search							#lists both formulae and casks
brew search --formulae				#lists only formulae
brew search --casks 				#lists only casks
brew search <arg>					#lists both formulae and casks
brew search --formulae <arg>		#lists only formulae
brew search --casks <arg>			#lists only casks
```

I've also decided to add tests in a separate PR (if that's okay with you?) and I'll be working on it soon. The changes have been tested locally on my system as well.

Please let me know if you have any questions regarding the edge cases. 

Looking forward to your feedback!

Thank you. 😃

**Comment by MikeMcQuaid** [comment]
> brew search							#lists both formulae and casks

I think this is the bit that should only list formulae. Otherwise: sounds good!

File: Library/Homebrew/cmd/search.rb [hunk 1]
```diff
@@ -64,7 +67,9 @@ def search
     end
 
     if args.remaining.empty?
-      if args.casks?
+      if args.formulae?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]: Might want to think here what it should do if `--formulae` and `--casks` is passed? I'd suggest a `raise UsageError, "Cannot specify --formulae and --casks without <text>!`

File: Library/Homebrew/cmd/search.rb [hunk 2]
```diff
@@ -89,15 +94,28 @@ def search
       remote_casks = remote_results[:casks]
       all_casks = local_casks + remote_casks
 
-      if all_formulae.any?
-        ohai "Formulae"
-        puts Formatter.columns(all_formulae)
-      end
+      if args.formulae?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]: Similarly: what if `--formulae` and `--casks` are passed here? Probably want to output them both.

**Comment by vidusheeamoli** [comment]
> I think this is the bit that should only list formulae. Otherwise: sounds good!

Oh, this is already incorporated, only the comment is wrong. 😛 

Also, addressing rest of the reviews right now. Thank you! 😄

File: Library/Homebrew/cmd/search.rb (hunk 2 above)
Comment (vidusheeamoli) [reply Library/Homebrew/cmd/search.rb]: Yes, makes sense!

File: Library/Homebrew/cmd/search.rb (hunk 1 above)
Comment (vidusheeamoli) [reply Library/Homebrew/cmd/search.rb]: On it.

**Comment by vidusheeamoli** [comment]
Sorry for the delay, a brew style test had failed earlier. Thank you for your patience @MikeMcQuaid! 😃

File: Library/Homebrew/cmd/search.rb [hunk 3]
```diff
@@ -63,6 +66,9 @@ def search
       return
     end
 
+    formulae_and_casks_present = args.formulae? && args.casks?
+    raise UsageError, "Cannot specify --formulae and --casks without an argument!" if formulae_and_casks_present
+
     if args.remaining.empty?
       if args.casks?
         puts Formatter.columns(Cask::Cask.to_a.map(&:full_name).sort)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]:
```suggestion
    if args.remaining.empty?
      if args.casks?
        raise UsageError, "Cannot specify --formulae and --casks without an argument!" if args.formulae?
        puts Formatter.columns(Cask::Cask.to_a.map(&:full_name).sort)
```

File: Library/Homebrew/cmd/search.rb [hunk 4]
```diff
@@ -89,15 +95,28 @@ def search
...
+        if all_casks.any?
+          puts if all_formulae.any?
+          ohai "Casks"
+          puts Formatter.columns(all_casks)
+        end
+      else
+        if all_formulae.any?
+          ohai "Formulae"
+          puts Formatter.columns(all_formulae)
+        end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]:
```suggestion
      if args.formulae? && all_formulae.any?
        ohai "Formulae"
        puts Formatter.columns(all_formulae)
      end
      
      if args.casks? && all_casks.any?
        puts if args.formulae? && all_formulae.any?
        ohai "Casks"
        puts Formatter.columns(all_casks)
      end
```

File: Library/Homebrew/cmd/search.rb [hunk 5]
```diff
@@ -89,15 +95,28 @@ def search
...
-      if all_casks.any?
-        puts if all_formulae.any?
-        ohai "Casks"
-        puts Formatter.columns(all_casks)
+        if all_casks.any?
+          puts if all_formulae.any?
+          ohai "Casks"
+          puts Formatter.columns(all_casks)
+        end
       end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]:
```suggestion
```

File: Library/Homebrew/cmd/search.rb [hunk 6]
```diff
@@ -63,6 +66,9 @@ def search
       return
     end
 
+    formulae_and_casks_present = args.formulae? && args.casks?
+    raise UsageError, "Cannot specify --formulae and --casks without an argument!" if formulae_and_casks_present
+
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]:
```suggestion
```

File: Library/Homebrew/cmd/search.rb (hunk 5 above)
Comment (vidusheeamoli) [reply Library/Homebrew/cmd/search.rb]: Thank you for the feedback @MikeMcQuaid! I understand the reason behind all these changes, but I'd like to point out that the following case is being missed:
`brew search <argument>`

I think that we'll have to add another check for cases where no flags are passed, and display both casks and formulae in such cases.

File: Library/Homebrew/cmd/search.rb [hunk 7]
```diff
@@ -89,17 +94,30 @@ def search
       remote_casks = remote_results[:casks]
       all_casks = local_casks + remote_casks
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]:
```suggestion
	       print_formulae = args.formulae?
	       print_casks = args.casks?
	       print_formulae = print_casks = true if !print_formulae && !print_casks
```

File: Library/Homebrew/cmd/search.rb [hunk 8]
```diff
@@ -89,17 +94,30 @@ def search
...
 
-      if all_formulae.any?
+      if args.formulae? && all_formulae.any?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]:
```suggestion
      if print_formulae && all_formulae.any?
```

File: Library/Homebrew/cmd/search.rb [hunk 9]
```diff
@@ -89,17 +94,30 @@ def search
...
         ohai "Formulae"
         puts Formatter.columns(all_formulae)
       end
 
-      if all_casks.any?
-        puts if all_formulae.any?
+      if args.casks? && all_casks.any?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]:
```suggestion
      if print_casks && all_casks.any?
```

File: Library/Homebrew/cmd/search.rb [hunk 10]
```diff
@@ -89,17 +94,30 @@ def search
...
+          puts Formatter.columns(all_formulae)
+        end
+
+        if all_casks.any?
+          puts if args.formulae? && all_formulae.any?
+          ohai "Casks"
+          puts Formatter.columns(all_casks)
+        end
+      end
+
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/search.rb]:
```suggestion
```

File: Library/Homebrew/cmd/search.rb (hunk 5 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/search.rb]: Good point!

**Comment by vidusheeamoli** [comment]
@BrewTestBot test this please

**Comment by EricFromCanada** [comment]
Resolves #6761.
